Mathlib/NumberTheory/ModularForms/Petersson.lean

```diff
@@ -0,0 +1,82 @@
+`Γ`.
+-/
+
+
+open UpperHalfPlane
+
+open scoped MatrixGroups ComplexConjugate ModularForm
+
+namespace UpperHalfPlane
+
+/-- The integrand in the Petersson scalar product of two modular forms. -/
+noncomputable def petersson (k : ℤ) (f f' : ℍ → ℂ) (τ : ℍ) :=
+  conj (f τ) * f' τ * τ.im ^ k
+
+lemma petersson_continuous (k : ℤ) {f f' : ℍ → ℂ} (hf : Continuous f) (hf' : Continuous f') :
+    Continuous (petersson k f f') := by
+  apply ((Complex.continuous_conj.comp hf).mul hf').mul
+  apply (Complex.continuous_ofReal.comp continuous_im).zpow₀
+  exact fun τ ↦ .inl <| Complex.ofReal_ne_zero.mpr τ.im_ne_zero
+
+lemma petersson_slash (k : ℤ) (f f' : ℍ → ℂ) (g : GL (Fin 2) ℝ) (τ : ℍ) :
+    petersson k (f ∣[k] g) (f' ∣[k] g) τ =
+      |g.det.val| ^ (k - 2) * σ g (petersson k f f' (g • τ)) := by
+  set D := |g.det.val|
+  have hD : (D : ℂ) ≠ 0 := Complex.ofReal_ne_zero.mpr <| abs_ne_zero.mpr <| g.det_ne_zero
+  set j := denom g τ
+  calc petersson k (f ∣[k] g) (f' ∣[k] g) τ
+  _ = D ^ (2 * k - 2) * conj (σ g (f (g • τ))) * σ g (f' (g • τ))
+      * (τ.im ^ k * j.normSq ^ (-k)) := by
+    simp [Complex.normSq_eq_conj_mul_self, (by ring : 2 * k - 2 = (k - 1) + (k - 1)), petersson,
+      zpow_add₀ hD, mul_zpow, ModularForm.slash_def, -Matrix.GeneralLinearGroup.val_det_apply]
+    ring
+  _ = D ^ (k - 2) * (conj (σ g (f (g • τ))) * σ g (f' (g • τ))
+      * (D * τ.im / j.normSq) ^ k) := by
+    rw [div_zpow, mul_zpow, zpow_neg, div_eq_mul_inv, (by ring : 2 * k - 2 = k + (k - 2)),
+      zpow_add₀ hD]
+    ring
+  _ = D ^ (k - 2) * (conj (σ g (f (g • τ))) * σ g (f' (g • τ)) * (im (g • τ)) ^ k) := by
+    rw [im_smul_eq_div_normSq, Complex.ofReal_div, Complex.ofReal_mul]
+  _ = D ^ (k - 2) * σ g (petersson k f f' (g • τ)) := by simp [petersson, σ_conj]
+
+lemma petersson_slash_SL (k : ℤ) (f f' : ℍ → ℂ) (g : SL(2, ℤ)) (τ : ℍ) :
+    petersson k (f ∣[k] g) (f' ∣[k] g) τ = petersson k f f' (g • τ) := by
+  -- need to disable two simp lemmas as they work against `Matrix.SpecialLinearGroup.det_coe`
+  simp [σ, ModularForm.SL_slash, petersson_slash,
+    -Matrix.SpecialLinearGroup.map_apply_coe, -Matrix.SpecialLinearGroup.coe_matrix_coe]
+
+end UpperHalfPlane
+
+section
+
+variable {F F' : Type*} [FunLike F ℍ ℂ] [FunLike F' ℍ ℂ]
+
+lemma SlashInvariantFormClass.petersson_smul {k : ℤ} {Γ : Subgroup SL(2, ℤ)}
+    [SlashInvariantFormClass F Γ k] {f : F} [SlashInvariantFormClass F' Γ k] {f' : F'}
+    {g : SL(2, ℤ)} (hg : g ∈ Γ) {τ : ℍ} :
+    petersson k f f' (g • τ) = petersson k f f' τ := by
+  simpa only [SlashInvariantFormClass.slash_action_eq _ _ hg]
+    using (petersson_slash_SL k f f' g τ).symm
+
+lemma ModularFormClass.petersson_continuous (k : ℤ) (Γ : Subgroup SL(2, ℤ))
+    [ModularFormClass F Γ k] [ModularFormClass F' Γ k] (f : F) (f' : F') :
+    Continuous (petersson k f f') :=
+  UpperHalfPlane.petersson_continuous k
+    (ModularFormClass.holo f).continuous (ModularFormClass.holo f').continuous
```

Comment:
`fun_prop` should be able to do that, but this will require some work.
Or maybe like this:
```lean
@[fun_prop]
lemma ModularFormClass.continuous (k : ℤ) (Γ : Subgroup SL(2, ℤ)) [ModularFormClass F Γ k] (f : F) :
  Continuous f := (ModularFormClass.holo f).continuous

lemma ModularFormClass.petersson_continuous (k : ℤ) (Γ : Subgroup SL(2, ℤ))
    [ModularFormClass F Γ k] [ModularFormClass F' Γ k] (f : F) (f' : F') :
    Continuous (petersson k f f') := by
  fun_prop
```
(Possibly with the additional `fun_prop` lemmas, this lemma is not even necessary.)